runtime/wasm/src/module/mod.rs

```diff
@@ -591,17 +618,38 @@ impl AscHeap for WasmInstanceContext {
         self.arena_start_ptr += size;
         self.arena_free_size -= size;
 
-        ptr as u32
+        Ok(ptr as u32)
     }
 
-    fn get(&self, offset: u32, size: u32) -> Vec<u8> {
+    fn get(&self, offset: u32, size: u32) -> Result<Vec<u8>, DeterministicHostError> {
         let offset = offset as usize;
         let size = size as usize;
 
+        let end = offset.checked_add(size).ok_or_else(|| {
+            DeterministicHostError(anyhow!(
+                "Overflow when accessing heap slice. Offset: {} Size: {}",
+                offset,
+                size
+            ))
+        })?;
+
         // Safety:
         // This unsafe block has been checked to not cause unsoundness by itself.
         // See 2155cdca-dfaa-4fba-86e4-289e7683c1bf for why this is sufficient.
-        unsafe { self.memory.data_unchecked()[offset..(offset + size)].to_vec() }
+        let data = unsafe {
+            self.memory
+                .data_unchecked()
+                .get(offset..end)
+                .map(|s| s.to_vec())
+        };
```

Comment:
We should be able to get rid of these `unsafe`s soon https://github.com/bytecodealliance/wasmtime/pull/2528.